### ISSUE TITLE
Redundant usage of Boolean.not()

### ISSUE THREAD

**Issue by osipxd** [detekt/detekt#7019]
***Obsolete.** Superseded by #7034 and #7035*

## Expected Behavior of the rule

Suggest to replace `Boolean.not()` by the operator `!` if it is possible.

```kotlin
// Compliant
if (!isValid) { ... } 
val notValid = isValid?.not() // OK to use .not() for nullable types 

// Noncompliant
if (isValid.not()) { ... } // Call .not() could be replaced by the operator !
```

## Context

- It might confuse if both of options  are used for boolean iversion. It is harder to understand if a condition is negative because you have to check both the start and the end of the condition.
- Automatic refactorings like "Invert 'If' condition" doesn't work correctly with `.not()`. Applying of such refactoring produces double-negative expression `!isValid.not()`

**Comment by atulgpt** [comment]
Hi @osipxd I think you can use the `ForbiddenMethod` rule for this thing. Are you not able to do that?

**Comment by osipxd** [comment]
> I think you can use the `ForbiddenMethod` rule for this thing.

Unfortunately not. When I add pattern `kotlin.Boolean.not` to `ForbiddenMethod`, detekt reports also `!`, because `.not()` is `operator fun`.

**Comment by osipxd** [comment]
However, this rule might be generalized to "Redundant explicit usage of an operator" with configurable list of operators to be reported.
```yaml
ExplicitOperatorCall:
    enabled: true
    operators:
      - not
      - plusAssign
```

**Comment by osipxd** [comment]
By the way, @atulgpt. I've noticed that you prefer to use `.not()` (at least in this repository), can you elaborate on why should it be used instead of `!`?

I know the point that `!` is easy to miss in expression, but I'm not sure usage of explicit operator completely solves this problem. In addition to the reasons I've described in "Context" section, there is another one reason: until `!` is not prohibited from usage, we cannot ensure there will no construction like `!isValid.not()`. This problem could be solved with something like "Double negative expression" rule, by the way.

<details>
<summary>Some statistics</summary>

[Usages of `!` in detekt project](https://github.com/search?q=repo%3Adetekt%2Fdetekt+%21+AND+NOT+%2F%28%21%3D%7C%21%21%7C%21is+%7C%21in+%7C%22.*%21.*%22%29%2F+language%3AKotlin&type=code&p=4) ~70 files
[Usages of `.not()` in detekt project](https://github.com/search?q=repo%3Adetekt%2Fdetekt+.not%28%29+language%3AKotlin&type=code) ~18 files (2 of them is usage of `.not()` on a nullable type)

</details>

**Comment by atulgpt** [comment]
> By the way, @atulgpt. I've noticed that you prefer to use `.not()` (at least in this repository), can you elaborate on why should it be used instead of `!`?
> 
> I know the point that `!` is easy to miss in expression, but I'm not sure that usage of explicit operator completely solves this problem. In addition to the reasons I've described in "Context" section, there is another one reason: until `!` is not prohibited from usage, we cannot ensure there will no construction like `!isValid.not()`. However, this problem could be solved with something like "Double negative expression" rule.
> 
> <details>
> <summary>Some statistics</summary>
> 
> [Usages of `!` in detekt project](https://github.com/search?q=repo%3Adetekt%2Fdetekt+%21+AND+NOT+%2F%28%21%3D%7C%21%21%7C%21is+%7C%21in+%7C%22.*%21.*%22%29%2F+language%3AKotlin&type=code&p=4) ~70 files
> [Usages of `.not()` in detekt project](https://github.com/search?q=repo%3Adetekt%2Fdetekt+.not%28%29+language%3AKotlin&type=code) ~18 files (2 of them is usage of `.not()` on a nullable type)
> 
> </details>

Hi @osipxd, I think it's my preference as I somehow find `.not` very easy to use as it  goes very well with my flow of thinking(as you have to apply it at the very end so I don't have to traverse the cursor back to add `! `) also it makes the expression easy to use in a chain of operators

But I am not against forbidding the use of `.not` in certain condition(it would have been awesome if we can accomplish this using ForbiddenMethod). It's just that I might not be using that rule in my projects

I completely agree with double Inverted rule as that pattern should be disallowed

**Comment by BraisGabin** [comment]
"Double negative expression" sounds like a great rule for detekt.

And about this rule I don't think we should have a rule that forbids the usage of `.not`. In general the people see detekt rules as something they should follow as good patterns. Even if they are not enable by default. So I think that your proposal of `RedundantExplicitOperatorCall` would be a better fit. Also we already have a rule to promote `[]` over `get` so maybe we can merge it (I can't recall the name now).

**Comment by osipxd** [comment]
Closing this one in favor of the created issues. Let's continue discussion there.

**Comment by osipxd** [comment]
Closing this one in favor of the created issues. Let's continue discussion there.
